csrc/abstract_tensor_schedule.cpp

```diff
@@ -0,0 +1,261 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+
+#include <abstract_tensor.h>
+#include <abstract_tensor_schedule.h>
+#include <ir/internal_base_nodes.h>
+#include <iter_visitor.h>
+#include <val_graph.h>
+
+namespace nvfuser {
+
+namespace {
+
+class AbstractTensorSchedule {
+ public:
+  static void apply(
+      const AbstractTensor& abstract,
+      TensorView* concrete,
+      ValGraph* graph) {
+    AbstractTensorSchedule ats(abstract, concrete, graph);
+    ats.run();
+  }
+
+ private:
+  AbstractTensorSchedule(
+      const AbstractTensor& abstract,
+      TensorView* concrete,
+      ValGraph* graph)
+      : abstract_(abstract), concrete_(concrete), graph_(graph) {}
+
+  void run() {
+    findNearestProducers();
+
+    // Now, for each ValGroup in abstract, find the closest producer ValGroups
+    // with entries in concrete_ids_ and replay the path from them. If none
+    // exists, then do not include this dimension in the output
+    std::vector<IterDomain*> loop_domain;
+    for (const AbstractId& abs_id : abstract_.domain) {
+      IterDomain* new_id = replayAbstractId(abs_id);
+      if (new_id == nullptr) {
+        continue;
+      }
+      loop_domain.push_back(new_id);
+    }
+    concrete_->setLoopDomain(loop_domain);
+  }
+
+  ValGroup abstractIdToValGroup(const AbstractId& abs_id) {
+    if (abs_id.is<IterDomain*>()) {
+      return graph_->toGroup(abs_id.as<IterDomain*>());
+    } else if (abs_id.is<ValGroupAndItsGraph>()) {
+      if (graph_ == nullptr) {
+        graph_ = abs_id.as<ValGroupAndItsGraph>().graph;
+      } else {
+        NVF_ERROR(graph_ == abs_id.as<ValGroupAndItsGraph>().graph);
+      }
+      return abs_id.as<ValGroupAndItsGraph>().group;
+    }
+    NVF_ERROR(false, "AbstractId must be IterDomain* or ValGroupAndItsGraph");
+  }
+
+  //! Work backward from each loop IterDomain in concrete. When we find an
+  //! IterDomain* with a ValGroup in scheduled_val_groups, we map to it
+  //! (including siblings) in concrete_ids_. Assert that we do find a
+  //! scheduled ValGroup, otherwise we would not know where to place this
+  //! IterDomain.
+  void findNearestProducers() {
```

Comment:
Could you provide an example in the comment?

Reply:
Added